tests/integration/azure/test_azure_credentials.py

```diff
@@ -0,0 +1,142 @@
+            variables["installation"],
+            ws,
+            variables["azure_resource_permissions"],
+            variables["azure_sp_crawler"],
+            integration_test_flag=integration_test_flag,
+        )
+        spn_migration.execute_migration(
+            MockPrompts({"Above Azure Service Principals will be migrated to UC storage credentials *": "Yes"})
+        )
+        return spn_migration
+
+    return inner
+
+
+def test_spn_migration_existed_storage_credential(
+    ws, execute_migration, make_storage_credential_from_spn, prepare_spn_migration_test
+):
+    variables = prepare_spn_migration_test(read_only=False)
+
+    # create a storage credential for this test
+    make_storage_credential_from_spn(
+        name=variables["storage_credential_name"],
+        application_id=variables["application_id"],
+        client_secret=variables["client_secret"],
+        directory_id=variables["directory_id"],
+    )
+
+    # test that the spn migration will be skipped due to above storage credential is existed
+    spn_migration = execute_migration(variables, integration_test_flag=variables["storage_credential_name"])
+
+    # because storage_credential is existing, no spn should be migrated
+    assert not spn_migration._final_sp_list
+
+
+@pytest.mark.parametrize("read_only", [False, True])
+def test_spn_migration(ws, execute_migration, prepare_spn_migration_test, read_only):
+    variables = prepare_spn_migration_test(read_only)
+
+    try:
+        spn_migration = execute_migration(variables, integration_test_flag="lets_migrate_the_spn")
+
+        assert spn_migration._final_sp_list[0].service_principal.principal == variables["storage_credential_name"]
+        assert ws.storage_credentials.get(variables["storage_credential_name"]).read_only is read_only
+
+        validation_result = spn_migration._installation.save.call_args.args[0][0]
+        if read_only:
+            # We only assert that write validation are not performed for read only storage credential here.
+            # In real life, the READ validation for read only storage credential may fail if there is no file,
+            # but that is fine, as the storage credential is created, and we just cannot validate it until it's really used.
+            assert not any(
+                (res.operation is not None) and ("WRITE" in res.operation.value) for res in validation_result.results
+            )
+        else:
+            assert any(
+                (res.operation is not None) and ("WRITE" in res.operation.value) and ("PASS" in res.result.value)
+                for res in validation_result.results
+            )
+            assert any(
+                (res.operation is not None) and ("DELETE" in res.operation.value) and ("PASS" in res.result.value)
+                for res in validation_result.results
+            )
+    finally:
+        ws.storage_credentials.delete(name=variables["storage_credential_name"], force=True)
```

Comment:
`name` is a positional argument! why do you use it as keyed?

https://databricks-sdk-py.readthedocs.io/en/latest/workspace/catalog/storage_credentials.html#databricks.sdk.service.catalog.StorageCredentialsAPI.create

Reply:
fixed